src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,956 @@
+                            foreach ((BoundPattern rewrittenPattern, SyntaxNode? syntax) in expandedLeftSet.Cases)
+                            {
+                                resultOrSet.Add(rewrittenPattern, syntax);
+                            }
+                        }
+                    }
+
+                    using SetsOfOrCases rightSetOfOrCases = RewriteToSetsOfOrCases(binaryPattern.Right);
+                    if (!rightSetOfOrCases.IsDefault)
+                    {
+                        // In `A and B`, when found multiple `or` patterns in `B` to be expanded
+                        // For each such nested expansion, we'll produce a `A and ...` expansion
+                        foreach (OrCases expandedRightSet in rightSetOfOrCases.Set)
+                        {
+                            OrCases resultOrSet = result.StartNewOrCases();
+                            foreach ((BoundPattern rewrittenPattern, SyntaxNode? syntax) in expandedRightSet.Cases)
+                            {
+                                resultOrSet.Add(binaryPattern.WithRight(rewrittenPattern), syntax);
+                            }
+                        }
+                    }
+                }
+
+                return result;
+            }
+
+            static SetsOfOrCases rewriteRecursive(BoundRecursivePattern recursivePattern)
+            {
+                SetsOfOrCases result = default;
+
+                // If any of the nested property sub-patterns can be expanded, we carry those on.
+                // For example, in `{ Prop: A, ... }`, when we found multiple `or` patterns in `A` to be expanded
+                // For each such nested expansion, we'll produce an `{ Prop: <expansion>, ... }` expansion
+                ImmutableArray<BoundPropertySubpattern> propertySubpatterns = recursivePattern.Properties;
+                if (!propertySubpatterns.IsDefault)
+                {
+                    for (int i = 0; i < propertySubpatterns.Length; i++)
+                    {
+                        BoundPattern pattern = propertySubpatterns[i].Pattern;
+                        using SetsOfOrCases setOfOrCases = RewriteToSetsOfOrCases(pattern);
+                        if (!setOfOrCases.IsDefault)
+                        {
+                            foreach (OrCases orSet in setOfOrCases.Set)
+                            {
+                                OrCases resultOrSet = result.StartNewOrCases();
+                                foreach ((BoundPattern resultPattern, SyntaxNode? syntax) in orSet.Cases)
+                                {
+                                    ImmutableArray<BoundPropertySubpattern> properties = recursivePattern.Properties;
+                                    BoundRecursivePattern resultRecursivePattern = recursivePattern
+                                        .WithProperties(properties.SetItem(i, properties[i].WithPattern(resultPattern)));
```

Comment:
>.WithProperties(properties.SetItem(i, properties[i].WithPattern(resultPattern))); [](http://example.com/codeflow?start=40&length=81)

Similarly here, I think there is no good reason to keep the tail (similar to how ```rewriteBinary``` cuts the tail for `and`). Again I would prefer ```MoveNotPatternsDownRewriter``` to do complete normalization. In that case we could safely assume that there is only one property sub-pattern in the recursive pattern and the logic here could be much simpler.